# Homework Assignment 2: Text Representation and Basic Text Mining

| Done by: Kevin Sánchez

## Part 1: Basic Concepts of Text Mining
---

**Define the following concepts in your own words**

1. **Corpus**: Is a large and structured collection of elements (e.g. words) that could be use for natural processing systems. In the context of data mining, corpus is use for generally analyze data.

2. **Document**: Is a single unit of text, which belongs to an entire dataset (corpus).

3. **Tokenization**: Is the process of breaking text into smaller units called tokens words. Depende of the techique the division following a different criteria. The breaking could be on: words, subwords, characters, sentences, symbols. The objetive of this techique is for computers can understand the text by: count words, analyze meaning, computer statistics, etc.

3. **Stopwords**: Are words that are not useful for training the models. Because they're little meaning for many text mining tasks helping reduce noise. Named stop-words because NLP systems often “stop” paying attention to them during preprocessing.

3. **Stemming vs Lemmatization**: Stemming and Lemmatization area two techniques used to reduce words to a simpler or base form. This technique help to the NLP systems for reduce vocabulary size, improve matching and group similar meanings. Stemming reduce word based on prefixes and suffixes using simple rules. Lemmatization in their side use a lemma, based on vocabulary, grammar, POS (part-of-speech) analysis. The more accurate technique is lemmatization, but the faster is stemming.

**Explain why text must be transformed into numerical representations before applying machine
learning algorithms.**

Because computers only understand mathematics, for example machine learning methods requires numeric inputs to do operations such as: addition or multiplication, distance calculations, probability estimation, gradient optimization, etc.. We have to convert text into numbers for process as vectors, matrices, to detect patterns, compute similarities, and optimize predictions.

**Describe at least three common preprocessing steps in text mining and explain their purpose**

1. Tokenization is the process of breaking text into smaller units called tokens, usually words or sentences.
2. Stop words removal. Words often carry little meaningful information.
3. Stemming or Lemmatization for reduce words to their root or base form.

## Part 2: Bag of Words Representation
---

Consider the following corpus of 3 documents

In [1]:
Document_1 = "I love natural language processing"
Document_2 = "Natural language processing is fun"
Document_3 = "I love machine learning"

Perform tokenization and convert all words to lowercase

In [2]:
from nltk.tokenize import word_tokenize

Document_1_tokens = word_tokenize(Document_1.lower())
Document_2_tokens = word_tokenize(Document_2.lower())
Document_3_tokens = word_tokenize(Document_3.lower())

print("Document 1 Tokens:", Document_1_tokens)
print("Document 2 Tokens:", Document_2_tokens)
print("Document 3 Tokens:", Document_3_tokens)

Document 1 Tokens: ['i', 'love', 'natural', 'language', 'processing']
Document 2 Tokens: ['natural', 'language', 'processing', 'is', 'fun']
Document 3 Tokens: ['i', 'love', 'machine', 'learning']


Build the vocabulary of the corpus

In [3]:
corpus = [item for lst in [Document_1_tokens, Document_2_tokens, Document_3_tokens] for item in lst]
corpus = set(corpus)
print("Corpus:", corpus)

Corpus: {'fun', 'natural', 'i', 'learning', 'language', 'is', 'love', 'machine', 'processing'}


Construct the Bag-of-Words (BoW) representation as a term-document matrix

In [4]:
from collections import defaultdict

def make_Bow(Corpus, Document):
    BoW = defaultdict(int)
    for element in Corpus:
        BoW[element]
        if element in Document:
            BoW[element] += 1
    return dict(BoW)

In [5]:
BoW_Document_1 = make_Bow(corpus, Document_1_tokens)

print("BoW Document 1:", BoW_Document_1)

BoW_Document_2 = make_Bow(corpus, Document_2_tokens)

print("BoW Document 2:", BoW_Document_2)

BoW_Document_3 = make_Bow(corpus, Document_3_tokens)
print("BoW Document 3:", BoW_Document_3)

BoW Document 1: {'fun': 0, 'natural': 1, 'i': 1, 'learning': 0, 'language': 1, 'is': 0, 'love': 1, 'machine': 0, 'processing': 1}
BoW Document 2: {'fun': 1, 'natural': 1, 'i': 0, 'learning': 0, 'language': 1, 'is': 1, 'love': 0, 'machine': 0, 'processing': 1}
BoW Document 3: {'fun': 0, 'natural': 0, 'i': 1, 'learning': 1, 'language': 0, 'is': 0, 'love': 1, 'machine': 1, 'processing': 0}


**What are the main limitations of the Bag-of-Words model?**
BoW have some limitations because of are a basic tecnhique to count the ocurrences of the words. Some limitations are the following:
1. BoWs trats a document as an unordered collection of words
2. It cannot undestand meaning, sarcasm, or relationships between words
3. Make BoWs, creeate large matrices that consume memory and computational resources
4. BoW ignores sentence structure, tense, negation and grammar
5. Small changes in wording create different features, such as "run", "running" and "ran"

## Part 3: Term Frequency (TF)
---

**Define Term Frequency (TF)**

Term Frequency is a measure used in text mining and information retrieval to indicate how often a word appears in a document. It helps determine the importance of a word within specific document.

The basic formula is:

$TF(t,d) = \frac{Number\ of\ times\ term\ t\ appears\ in\ document\ d}{Total\ number\ of\ terms\ in\ document\ d}$

Where:

$t = the\ term (word)$

$d = the\ document$

Compute the raw term frequency for each term in each document from Part 2

In [6]:
def compute_TF(Document, BoW):
    return dict(BoW)

In [7]:
TF_Document_1 = compute_TF(Document_1_tokens, BoW_Document_1)
print("TF Document 1:", TF_Document_1)

TF_Document_2 = compute_TF(Document_2_tokens, BoW_Document_2)
print("TF Document 2:", TF_Document_2)

TF_Document_3 = compute_TF(Document_3_tokens, BoW_Document_3)
print("TF Document 3:", TF_Document_3)

TF Document 1: {'fun': 0, 'natural': 1, 'i': 1, 'learning': 0, 'language': 1, 'is': 0, 'love': 1, 'machine': 0, 'processing': 1}
TF Document 2: {'fun': 1, 'natural': 1, 'i': 0, 'learning': 0, 'language': 1, 'is': 1, 'love': 0, 'machine': 0, 'processing': 1}
TF Document 3: {'fun': 0, 'natural': 0, 'i': 1, 'learning': 1, 'language': 0, 'is': 0, 'love': 1, 'machine': 1, 'processing': 0}


**Compute the normalized TF for each term using**:

$TF(t,d) = \frac{f_{t,d}}{\sum{t'}f_{t',d}}$

where $f_{t,d}$ is the frequency of term $t$ in document $d$

In [8]:
def compute_TF_normalized(Document, BoW):
    TF_dict = defaultdict(int)
    number_total_terms = len(Document)
    
    for term, frequency in BoW.items():
        TF_dict[term] = frequency / number_total_terms
    return dict(TF_dict)

In [9]:
TF_normalized_Document_1 = compute_TF_normalized(Document_1_tokens, BoW_Document_1)

print("TF Document 1:", TF_normalized_Document_1)

TF_normalized_Document_2 = compute_TF_normalized(Document_2_tokens, BoW_Document_2)
print("TF Document 2:", TF_normalized_Document_2)

TF_normalized_Document_3 = compute_TF_normalized(Document_3_tokens, BoW_Document_3)
print("TF Document 3:", TF_normalized_Document_3)

TF Document 1: {'fun': 0.0, 'natural': 0.2, 'i': 0.2, 'learning': 0.0, 'language': 0.2, 'is': 0.0, 'love': 0.2, 'machine': 0.0, 'processing': 0.2}
TF Document 2: {'fun': 0.2, 'natural': 0.2, 'i': 0.0, 'learning': 0.0, 'language': 0.2, 'is': 0.2, 'love': 0.0, 'machine': 0.0, 'processing': 0.2}
TF Document 3: {'fun': 0.0, 'natural': 0.0, 'i': 0.25, 'learning': 0.25, 'language': 0.0, 'is': 0.0, 'love': 0.25, 'machine': 0.25, 'processing': 0.0}


**Explain why normalization is useful.**

We normalize Term Frequency because raw word counts alone are not comparable across documents of different lenghts

## Part 4: Inverse Document Frequency (IDF)
---

**Define Inverse Document Frequency (IDF)**

Inverse Document Frequency (IDF) is a measure used in Information Retrieval and Natural Language Processing to determine how important or unique a word is across a collection of documents. The intuition behind IDF is inform how much important is a word that appears in many documents.

- High IDF -> The word is rare and important
- Low IDF -> The word is common and less useful for distinguishing documents

The $log$ is used mainly to smooth  the efect of rarity. 

Without the log, the ratio
$\frac{N}{df(t)}$

Where:
- $t$ = The term (word)
- $N$ = Total number of documents
- $df(t)$ = Number of document containing the term $t$

For example:
- A term appearing in 1 document out of 1,000,000 gets:
$\frac{1,000,000}{1} = 1,000,000$
- A term appearing 10 documents gets:
$\frac{1,000,000}{10} = 100,000$

Those raw values are huge and dominate the scoring too aggressively

Example with $N = 1,000,000$

|Document Frequency|Raw Ratio|Log IDF (base 10)|
|------------------|---------|-----------------|
|1|1,000,000|6|
|10|100,000|5|
|1,000|1,0000|3|
|100,000|10|1|

**Compute the IDF for each term using:**

$IDF(t) = \log(\frac{N}{df_{t}})$

where:
- $N$ is the total number of documents
- $df_{t}$ is the number of documents containing term t$t$

In [10]:
Collection_Documents = [Document_1, Document_2, Document_3]
total_number_documents = len(Collection_Documents)

def df(term):
    count = 0
    for doc in Collection_Documents:
        if term in doc:
            count += 1
    return count

In [16]:
import numpy as np

def IDF(term):
    return np.log(total_number_documents / df(term))

In [17]:
for term in corpus:
    print(f"IDF of '{term}': {IDF(term)}")

IDF of 'fun': 1.0986122886681098
IDF of 'natural': 1.0986122886681098
IDF of 'i': 0.0
IDF of 'learning': 1.0986122886681098
IDF of 'language': 0.4054651081081644
IDF of 'is': 1.0986122886681098
IDF of 'love': 0.4054651081081644
IDF of 'machine': 1.0986122886681098
IDF of 'processing': 0.4054651081081644


**Which terms have the highest IDF? Why?**

The terms with highest IDF are: 
- *fun*
- *natural*
- *learning*
- *language*
- *is*
- *machine*

The main reason because are words with less frequency "1". Therefore have similarity importance for all words.

**What happens to the IDF of very common words?**

With very common words, the value of IDF decrease. These means that the term is a very common word, therefore have less importance respect documents collection.

## Part 5: TF-IDF Representation
---

**Define TF-IDF**

Term Frequency - Inverse Document Frequency is a statistical measure used in Information Retrieval and NLP to evaluate how important a word is to a document within a collection of documents (corpus). Combine Term Frequency (TF) and Inverse Document Frequency (IDF). TF-IDF is commonly used for:
- search engines
- document ranking
- keword extraction
- text classification
- recommendation systems
- document similarity

**Compute the $TF$-$IDF$ score for each term in each document:**

$TF$-$IDF(t,d) = TF(t,d) * IDF(t)$

In [19]:
TF_IDF_Document_1 = {term: TF_normalized_Document_1[term] * IDF(term) for term in corpus}

print("TF-IDF Document 1:", TF_IDF_Document_1)

TF_IDF_Document_2 = {term: TF_normalized_Document_2[term] * IDF(term) for term in corpus}
print("TF-IDF Document 2:", TF_IDF_Document_2)

TF_IDF_Document_3 = {term: TF_normalized_Document_3[term] * IDF(term) for term in corpus}
print("TF-IDF Document 3:", TF_IDF_Document_3)

TF-IDF Document 1: {'fun': 0.0, 'natural': 0.21972245773362198, 'i': 0.0, 'learning': 0.0, 'language': 0.08109302162163289, 'is': 0.0, 'love': 0.08109302162163289, 'machine': 0.0, 'processing': 0.08109302162163289}
TF-IDF Document 2: {'fun': 0.21972245773362198, 'natural': 0.21972245773362198, 'i': 0.0, 'learning': 0.0, 'language': 0.08109302162163289, 'is': 0.21972245773362198, 'love': 0.0, 'machine': 0.0, 'processing': 0.08109302162163289}
TF-IDF Document 3: {'fun': 0.0, 'natural': 0.0, 'i': 0.0, 'learning': 0.27465307216702745, 'language': 0.0, 'is': 0.0, 'love': 0.1013662770270411, 'machine': 0.27465307216702745, 'processing': 0.0}


**Construct the TF-IDF matrix for the corpus.**

In [21]:
TF_IDF_Document_1

{'fun': 0.0,
 'natural': 0.21972245773362198,
 'i': 0.0,
 'learning': 0.0,
 'language': 0.08109302162163289,
 'is': 0.0,
 'love': 0.08109302162163289,
 'machine': 0.0,
 'processing': 0.08109302162163289}

In [32]:
import pandas as pd

df_Document_1 = pd.DataFrame([TF_IDF_Document_1])
df_Document_2 = pd.DataFrame([TF_IDF_Document_2])
df_Document_3 = pd.DataFrame([TF_IDF_Document_3])

tfidf_matrix = pd.concat([df_Document_1, df_Document_2, df_Document_3])
tfidf_matrix.index = ['Document 1', 'Document 2', 'Document 3']

tfidf_matrix

,fun,natural,i,learning,language,is,love,machine,processing
Document 1,0.000000,0.219722,0.0,0.000000,0.081093,0.000000,0.081093,0.000000,0.081093
Document 2,0.219722,0.219722,0.0,0.000000,0.081093,0.219722,0.000000,0.000000,0.081093
Document 3,0.000000,0.000000,0.0,0.274653,0.000000,0.000000,0.101366,0.274653,0.000000


**Identify the most important term in each document according to TF-IDF.**

The most important term in each document are the following:
- *Document 1*: "natural" is an important word for Document 1.
- *Document 2*: "natural, fun, is" are important words for Document 2.
- *Document 3*: "learning, machine" are important words for Document 3.

**Explain intuitively why TF-IDF works better than raw frequency.**

TF-IDF works better because consider statistical approaches unlike raw frequency that consider only the word's ocurrences. TF-IDF use two previous processes. "Term Frequency" and "Inverse Document Frequency" giving more details about the importance of a word considering the corpus and frequency.

## Part 6: Practical Implementation
---

**Use TfidfVectorizer to compute TF-IDF for the corpus.**

In [37]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer()

tfidf_matrix = vectorizer.fit_transform(Collection_Documents)

feature_names = vectorizer.get_feature_names_out()

tfidf_vectorizer = pd.DataFrame(tfidf_matrix.toarray(), columns=feature_names, index=['Document_1','Document_2','Document_3'])

tfidf_vectorizer

,fun,is,language,learning,love,machine,natural,processing
Document_1,0.00000,0.00000,0.500000,0.000000,0.50000,0.000000,0.500000,0.500000
Document_2,0.51742,0.51742,0.393511,0.000000,0.00000,0.000000,0.393511,0.393511
Document_3,0.00000,0.00000,0.000000,0.622766,0.47363,0.622766,0.000000,0.000000


**Compare your manual results with the library output.**

*Result with computation of TF-IDF manual:*
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>fun</th>
      <th>natural</th>
      <th>i</th>
      <th>learning</th>
      <th>language</th>
      <th>is</th>
      <th>love</th>
      <th>machine</th>
      <th>processing</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Document 1</th>
      <td>0.000000</td>
      <td>0.219722</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.081093</td>
      <td>0.000000</td>
      <td>0.081093</td>
      <td>0.000000</td>
      <td>0.081093</td>
    </tr>
    <tr>
      <th>Document 2</th>
      <td>0.219722</td>
      <td>0.219722</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.081093</td>
      <td>0.219722</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.081093</td>
    </tr>
    <tr>
      <th>Document 3</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.274653</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.101366</td>
      <td>0.274653</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
</div>

*Result with computation of TF-IDF using scikit-learn:*
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>fun</th>
      <th>is</th>
      <th>language</th>
      <th>learning</th>
      <th>love</th>
      <th>machine</th>
      <th>natural</th>
      <th>processing</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Document_1</th>
      <td>0.00000</td>
      <td>0.00000</td>
      <td>0.500000</td>
      <td>0.000000</td>
      <td>0.50000</td>
      <td>0.000000</td>
      <td>0.500000</td>
      <td>0.500000</td>
    </tr>
    <tr>
      <th>Document_2</th>
      <td>0.51742</td>
      <td>0.51742</td>
      <td>0.393511</td>
      <td>0.000000</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.393511</td>
      <td>0.393511</td>
    </tr>
    <tr>
      <th>Document_3</th>
      <td>0.00000</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.622766</td>
      <td>0.47363</td>
      <td>0.622766</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
</div>

We have differences between the values, the main reason is that behind the implementation of `TfidVectorizer` use L2-Normalization on the process. However in term of scales the level of importance are similar in each word.

**Experiment with:**
- Removing stopwords
- Using n-grams (e.g., bigrams)

In [50]:
# Experiment 1: Remove stopwords
vectorizer_stopwords = TfidfVectorizer(stop_words='english')
tfidf_matrix_stopwords = vectorizer_stopwords.fit_transform(Collection_Documents)
feature_names_stopwords = vectorizer_stopwords.get_feature_names_out()
tfidf_vectorizer_stopwords = pd.DataFrame(tfidf_matrix_stopwords.toarray(), columns=feature_names_stopwords, index=['Document_1','Document_2','Document_3'])

tfidf_vectorizer_stopwords

,fun,language,learning,love,machine,natural,processing
Document_1,0.000000,0.500000,0.000000,0.50000,0.000000,0.500000,0.500000
Document_2,0.604652,0.459854,0.000000,0.00000,0.000000,0.459854,0.459854
Document_3,0.000000,0.000000,0.622766,0.47363,0.622766,0.000000,0.000000


In [45]:
# Experiment 2: Use bigrams (2-word sequences)
vectorizer_bigrams = TfidfVectorizer(ngram_range=(2, 2))
tfidf_matrix_bigrams = vectorizer_bigrams.fit_transform(Collection_Documents)
feature_names_bigrams = vectorizer_bigrams.get_feature_names_out()
tfidf_vectorizer_bigrams = pd.DataFrame(tfidf_matrix_bigrams.toarray(), columns=feature_names_bigrams, index=['Document_1','Document_2','Document_3'])

tfidf_vectorizer_bigrams

,is fun,language processing,love machine,love natural,machine learning,natural language,processing is
Document_1,0.000000,0.517856,0.000000,0.680919,0.000000,0.517856,0.000000
Document_2,0.562829,0.428046,0.000000,0.000000,0.000000,0.428046,0.562829
Document_3,0.000000,0.000000,0.707107,0.000000,0.707107,0.000000,0.000000


In [46]:
# Experiment 3: Both stopwords + bigrams
vectorizer_both = TfidfVectorizer(stop_words='english', ngram_range=(2, 2))
tfidf_matrix_both = vectorizer_both.fit_transform(Collection_Documents)
feature_names_both = vectorizer_both.get_feature_names_out()
tfidf_vectorizer_both = pd.DataFrame(tfidf_matrix_both.toarray(), columns=feature_names_both, index=['Document_1','Document_2','Document_3'])

tfidf_vectorizer_both

,language processing,love machine,love natural,machine learning,natural language,processing fun
Document_1,0.517856,0.000000,0.680919,0.000000,0.517856,0.000000
Document_2,0.517856,0.000000,0.000000,0.000000,0.517856,0.680919
Document_3,0.000000,0.707107,0.000000,0.707107,0.000000,0.000000


In [47]:
# Experiment 4: Unigrams + bigrams (1 to 2 words)
vectorizer_mixed = TfidfVectorizer(ngram_range=(1, 2))
tfidf_matrix_mixed = vectorizer_mixed.fit_transform(Collection_Documents)
feature_names_mixed = vectorizer_mixed.get_feature_names_out()
tfidf_vectorizer_mixed = pd.DataFrame(tfidf_matrix_mixed.toarray(), columns=feature_names_mixed, index=['Document_1','Document_2','Document_3'])

tfidf_vectorizer_mixed

,fun,is,is fun,language,language processing,learning,love,love machine,love natural,machine,machine learning,natural,natural language,processing,processing is
Document_1,0.000000,0.000000,0.000000,0.359700,0.359700,0.000000,0.359700,0.000000,0.472963,0.000000,0.000000,0.359700,0.359700,0.359700,0.000000
Document_2,0.380914,0.380914,0.380914,0.289695,0.289695,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.289695,0.289695,0.289695,0.380914
Document_3,0.000000,0.000000,0.000000,0.000000,0.000000,0.467351,0.355432,0.467351,0.000000,0.467351,0.467351,0.000000,0.000000,0.000000,0.000000


In [49]:
# Experiment 5: Both stopwords + Trigrams
vectorizer_trigrams = TfidfVectorizer(stop_words='english', ngram_range=(3, 3))
tfidf_matrix_trigrams = vectorizer_trigrams.fit_transform(Collection_Documents)
feature_names_trigrams = vectorizer_trigrams.get_feature_names_out()
tfidf_vectorizer_trigrams = pd.DataFrame(tfidf_matrix_trigrams.toarray(), columns=feature_names_trigrams, index=['Document_1','Document_2','Document_3'])

tfidf_vectorizer_trigrams

,language processing fun,love machine learning,love natural language,natural language processing
Document_1,0.000000,0.0,0.795961,0.605349
Document_2,0.795961,0.0,0.000000,0.605349
Document_3,0.000000,1.0,0.000000,0.000000


**Briefly discuss the impact of these changes**

1. The impact of remove stop words have influence in the reduction of computation to create the matrix table.
2. Using bigrams or Trigrams added with stop words growth the importance of text.
3. Experimenting on the union of unigrams, bigrams without remove stopword. Create a large matrix with a lot of values to analyze the TF-IDF of differente posibilities. 
We have more information about the documents, however the computational cost for making is higher.

## Part 7: Reflection
---

**What are the main limitations of TF-IDF?**

TF-IDF ignores completely the semantics (synonyms, analogies, concepts, etc) of word because of only count words. For example when we mention car or automobile, the meaning refers to the same thing. However TF-IDF treats as different words. Also for words with the same syntaxis, but with different semantics depends on the contex (e.g. bank for river or financual institution).

Ignores word order and conext because of use BoW approach, and sentences with similiar representations that contain the same word can't be identified. For example
- "dog bites man" and "man bites dog"

If we have more words per document, the computational efficiency is affected in terms of memory. Finally TF-IDF prioritize rare words as very importants (e.g. typos, noise, very uncommon names), and not understand the importance of the document based on their authority, credibility and user intent.

**In what scenarios might TF-IDF fail?**

TF-IDF can fail or perform poorly in several real-world scenarios, especially when understanding meaning, context, or complex language patterns is important. For example when texts using different word for similar meanings. A lot of the scenarios about limitations of TF-IDF was detailed in the above question. However, some specific scenarios are specified below:

1. *Conversational AI and Chatbots*: TF-IDF cannot understand (intent, dialogue context, reasoning) and previuse conversational context.
2. *Recommendation Systems*: TF-IDF only analyzes text overlap, it cannot learn about user preferences, behavorial patterns and latent interests. For example two movies could be
very similar however TF-IDF can't identify because each movie use different vocabulary.
3. *Complex Question Answering*: TF-IDF cannot reason or combine facts to do inference about text or questions.

**Mention one modern alternative to TF-IDF (e.g., word embeddings) and briefly explain it.**

One modern alternative to TF-IDF is word embeddings, such as Word2Vec. This new approach represent words as dense numerical vectors in a continuous vector space. Unlinke TF-IDF,
embeddings capture semantic meaning and relationships between words. For example we can capture the following example:

- "king" and "queen"
- "car" and "automobile"

will have similar vector representations because they appear in similar contexts. Thanks to embeddings models can inference that words withm similar contexts tend to have similar meanings. Thanks to this approach embedding models can understand:
- synonyms
- semantic similarity
- analogies
- contextual relationships

Word embeddings are widely used in:
- semantic search
- recommendation systems
- chatbots
- machine translation
- modern NLP systems